[RA2 Ch04] Restructuring 4.6 Storage Components (#1870)
cntt-n/CNTT: doc/ref_arch/kubernetes/chapters/chapter04.md

|Ref|Specification|Details|Requirement Trace|
|---|---|---|---|

> |`ra2.stg.001`| Ephemeral Storage | An implementation must support ephemeral storage, for the unpacked container images to be stored and executed from, as a directory in the filesystem on the worker node on which the container is running <br>See the [Container runtimes](#4.4) section above for more information how this meets the requirement for ephemeral storage for containers <br>Ephemeral storage exists for the life of a container<br>Kubernetes Volumes attached using a Volume plugin can also be used to deliver Ephemeral Storage (may require additional tasks to ensure dleetion of ephemral storge on teh shutting down of a container) ||

Reviewer: ```suggestion
|`ra2.stg.001`| Ephemeral Storage | An implementation must support ephemeral storage, for the unpacked container images to be stored and executed from, as a directory in the filesystem on the worker node on which the container is running <br>See the [Container runtimes](#4.4) section above for more information how this meets the requirement for ephemeral storage for containers <br>Ephemeral storage exists for the life of a container<br>Kubernetes Volumes attached using a Volume plugin can also be used to deliver Ephemeral Storage (may require additional tasks to ensure deletion of ephemeral storage on the shutting down of a container) ||
```

Author: @CsatariGergely Sorry, I don't see a suggested change. 

"on" is missing from "... section above for more information how this meets the requirement for ephemeral storage ..." ---> "... section above for more information on how this meets the requirement for ephemeral storage ..."

> |`ra2.stg.001`| Ephemeral Storage | An implementation must support ephemeral storage, for the unpacked container images to be stored and executed from, as a directory in the filesystem on the worker node on which the container is running <br>See the [Container runtimes](#4.4) section above for more information how this meets the requirement for ephemeral storage for containers <br>Ephemeral storage exists for the life of a container<br>Kubernetes Volumes attached using a Volume plugin can also be used to deliver Ephemeral Storage (may require additional tasks to ensure deletion of ephemeral storage on the shutting down of a container) ||

Author: ```suggestion
|`ra2.stg.001`| Ephemeral Storage | An implementation must support ephemeral storage, for the unpacked container images to be stored and executed from, as a directory in the filesystem on the worker node on which the container is running <br>See the [Container runtimes](#4.4) section above for more information on how this meets the requirement for ephemeral storage for containers <br>Ephemeral storage exists for the life of a container<br>Kubernetes Volumes attached using a Volume plugin can also be used to deliver Ephemeral Storage (may require additional tasks to ensure deletion of ephemeral storage on the shutting down of a container) ||
```

> |`ra2.stg.001`| Ephemeral Storage | An implementation must support ephemeral storage, for the unpacked container images to be stored and executed from, as a directory in the filesystem on the worker node on which the container is running <br>See the [Container runtimes](#4.4) section above for more information on how this meets the requirement for ephemeral storage for containers <br>Ephemeral storage exists for the life of a container<br>Kubernetes Volumes attached using a Volume plugin can also be used to deliver Ephemeral Storage (may require additional tasks to ensure deletion of ephemeral storage on the shutting down of a container) ||

Reviewer 2: I'd also argue the following is unnecessary here and should be part of the chapter03.md content:

> Ephemeral storage exists for the life of a container<br>Kubernetes Volumes attached using a Volume plugin can also be used to deliver Ephemeral Storage (may require additional tasks to ensure deletion of ephemeral storage on the shutting down of a container).

> |`ra2.stg.001`| Ephemeral Storage | An implementation must support ephemeral storage, for the unpacked container images to be stored and executed from, as a directory in the filesystem on the worker node on which the container is running <br>See the [Container runtimes](#4.4) section above for more information on how this meets the requirement for ephemeral storage for containers <br>Ephemeral storage exists for the life of a container<br>Kubernetes Volumes attached using a Volume plugin can also be used to deliver Ephemeral Storage (may require additional tasks to ensure deletion of ephemeral storage on the shutting down of a container) ||

Reviewer 3: Agree with Tom

> |`ra2.stg.001`| Ephemeral Storage | An implementation must support ephemeral storage, for the unpacked container images to be stored and executed from, as a directory in the filesystem on the worker node on which the container is running <br>See the [Container runtimes](#4.4) section above for more information on how this meets the requirement for ephemeral storage for containers <br>Ephemeral storage exists for the life of a container<br>Kubernetes Volumes attached using a Volume plugin can also be used to deliver Ephemeral Storage (may require additional tasks to ensure deletion of ephemeral storage on the shutting down of a container) ||

Author: @tomkivlin @xmulligan 

Deleted content that belongs to RA Chapter 3:

- from ra2.stg.001: "Ephemeral storage exists for the life of a container.<br>Kubernetes Volumes attached using a Volume plugin can also be used to deliver Ephemeral Storage (may require additional tasks to ensure deletion of ephemeral storage on the shutting down of a container)" 

- from from ra2.stg.002: "This additional storage, that is attached to a container using a Kubernetes Volume, does not live and die with the container but instead follows the lifecycle of the Pod that the container is a part of. " 

- from ra2.stg.005: "The implementation may utilise either an In-tree or Out-of-tree type of Volume plugin.<br>In-tree plugins are built, linked, compiled and shipped with the core Kubernetes binaries and, thus, require the core Kubernetes code to be changed whenever a new backend storage system needs to be supported.<br>Out-of-tree plugins allow new backend storage systems to be supported without any changes to the core Kubernetes code." 

- from ra2.stg.007: "Storage Classes allow a consumer of the Kubernetes platform to request Persistent Storage using a Persistent Volume Claim and for a Persistent Volume to be dynamically created based on the "class" that has been requested. This avoids having to grant `create`/`update`/`delete` permissions in RBAC to PersistentVolume resources, which are cluster-scoped rather than namespace-scoped (meaning an identity can manage all PVs or none)."

> |`ra2.stg.002`| Kubernetes Volumes | An implementation may attach additional storage to containers using Kubernetes Volumes. This additional storage, that is attached to a container using a Kubernetes Volume, does not live and die with the container but instead follows the lifecycle of the Pod that the container is a part of ||

Reviewer 2: I'd also argue that the following is unnecessary and should be part of the chapter03.md content:

> This additional storage, that is attached to a container using a Kubernetes Volume, does not live and die with the container but instead follows the lifecycle of the Pod that the container is a part of.

|`ra2.stg.003`| Kubernetes Volumes | An implementation may use Volume Plugins (see `ra2.stg.005` below) to allow the use of a storage protocol (e.g. iSCSI, NFS) or management API (e.g. Cinder, EBS) for the attaching and mounting of storage into a Pod. ||

> |`ra2.stg.004`| Persistent Volumes | An implementation may support Kubernetes Persistent Volumes (PV) to provide persistent storage for Pods (requirement 'req.inf.stg.01'.<br>Persistent Volumes exist independent of the lifecycle of containers and/or pods. |[req.inf.stg.01](https://github.com/cntt-n/CNTT/blob/master/doc/ref_arch/kubernetes/chapters/chapter02.md#23-kubernetes-architecture-requirements)|

Reviewer 2: Suggest change to relative link.
```suggestion
|`ra2.stg.004`| Persistent Volumes | An implementation may support Kubernetes Persistent Volumes (PV) to provide persistent storage for Pods (requirement 'req.inf.stg.01'.<br>Persistent Volumes exist independent of the lifecycle of containers and/or pods. |[req.inf.stg.01](./chapter02.md#23-kubernetes-architecture-requirements)|
```

Author: @tomkivlin for readdocs (I maybe wrong) but I think we need the absolute url

Reviewer 2: > @tomkivlin for readdocs, I maybe wrong, I think we need the absolute url

Ah OK my bad, I hadn't realised that change.  Will resolve and approve.

> |`ra2.stg.004`| Persisten Volumes | An implemenetation may suppost Kubernetes Persisten Volumes (PV) to provide persistent storage for Pods (requirement 'req.inf.stg.01'<br>Persistent Volumes exist independnet of the lifecycle of containers and/or pods  ||

Reviewer: ```suggestion
|`ra2.stg.004`| Persisten Volumes | An implemenetation may support Kubernetes Persistent Volumes (PV) to provide persistent storage for Pods (requirement 'req.inf.stg.01'<br>Persistent Volumes exist independent of the lifecycle of containers and/or pods  ||
```

> |`ra2.stg.004`| Persisten Volumes | An implemenetation may support Kubernetes Persistent Volumes (PV) to provide persistent storage for Pods (requirement 'req.inf.stg.01'<br>Persistent Volumes exist independent of the lifecycle of containers and/or pods  ||

Reviewer 2: Adding final full stop and typos.
```suggestion
|`ra2.stg.004`| Persistent Volumes | An implementation may support Kubernetes Persistent Volumes (PV) to provide persistent storage for Pods (requirement 'req.inf.stg.01'<br>Persistent Volumes exist independent of the lifecycle of containers and/or pods. ||
```

Reviewer 2: I'd also argue the following is unnecessary here and should be in the chapter03.md content:

> Persistent Volumes exist independent of the lifecycle of containers and/or pods.

|`ra2.stg.005`| Storage Extension | Volume plugins must allow for the use of a range of backend storage systems. ||

> |`ra2.stg.006`| Container Storage Interface (CSI) | An implementation may support the Container Storage Interface (CSI), an Out-of-tree plugin.<br>In order to support CSI, the  feature gates `CSIDriverRegistry` and `CSINodeInfo` must be enabled.<br>The implementation must use a CSI driver (a full list of CSI drivers can be found [here](https://kubernetes-csi.github.io/docs/drivers.html)). <br>An implementation may support ephemeral storage through a CSI-compatible volume plugin in which case the `CSIInlineVolume` feature gate must be enabled.<br>An implementation may support Persistent Volumes through a CSI-compatible volume plugin in which case  the `CSIPersistentVolume` feature gate must be enabled. | |

Reviewer 3: Should the CSI be MAY or MUST?

Reviewer 2: @xmulligan is the question which should it be because both are used in the content, or because you think it should/shouldn't be one or the other?

Reviewer 3: The second, I think CSI should be a must requirement. The wording is also ambiguous right now because it switches between may and must

Author: @xmulligan Can make CSI as a "must" but there is no requirement in Ch02 for this. The paragraph currently states
1. that CSI **may** be supported
2. but if CSI is supported then certain capabilities **must** be enabled 

Again, if there is consensus can make the CSI to be *must**

> |`ra2.stg.005`| Storage Extension | Volume plugins must allow for the use of a range of backend storage systems<br>The implementation may utilise either an In-tree or Out-of-tree type of Volume plugin<br>In-tree plugins are built, linked, compiled and shipped with the core Kubernetes binaries and, thus, require the core KUbernetes code to be changed whenever a new backend storage system needs to be supported<br>Out-of-tree plugins allow new backend storage systems to be supported without any changes to the core Kubernetes code ||

Reviewer 2: Adding final full stop plus typos.
```suggestion
|`ra2.stg.005`| Storage Extension | Volume plugins must allow for the use of a range of backend storage systems<br>The implementation may utilise either an In-tree or Out-of-tree type of Volume plugin<br>In-tree plugins are built, linked, compiled and shipped with the core Kubernetes binaries and, thus, require the core Kubernetes code to be changed whenever a new backend storage system needs to be supported<br>Out-of-tree plugins allow new backend storage systems to be supported without any changes to the core Kubernetes code. ||
```

Reviewer 2: I'd also argue the following is unnecessary and should be in the chapter03.md content:

> In-tree plugins are built, linked, compiled and shipped with the core Kubernetes binaries and, thus, require the core Kubernetes code to be changed whenever a new backend storage system needs to be supported<br>Out-of-tree plugins allow new backend storage systems to be supported without any changes to the core Kubernetes code.

> |`ra2.stg.006`| Container Storage Interface (CSI) | An implementation may support the Container Storage Interface (CSI), an Out-of-tree plugin<br>In order to support CSI, the  feature gates `CSIDriverRegistry` and `CSINodeInfo` must be enabled<br>The implementation must use a CSI driver (a full list of CSI drivers can be found [here](https://kubernetes-csi.github.io/docs/drivers.html)) <br>An implementation may support ephemeral storage  through a CSI-compatible volume plugin in which case the `CSIInlineVolume` feature gate must be enabled<br>An implementation may support Persistent Volumes through a CSI-compatible volume plugin in which case  the `CSIPersistentVolume` feature gate must be enabled  | |

Reviewer 2: Adding final full stop.
```suggestion
|`ra2.stg.006`| Container Storage Interface (CSI) | An implementation may support the Container Storage Interface (CSI), an Out-of-tree plugin.<br>In order to support CSI, the  feature gates `CSIDriverRegistry` and `CSINodeInfo` must be enabled.<br>The implementation must use a CSI driver (a full list of CSI drivers can be found [here](https://kubernetes-csi.github.io/docs/drivers.html)). <br>An implementation may support ephemeral storage through a CSI-compatible volume plugin in which case the `CSIInlineVolume` feature gate must be enabled.<br>An implementation may support Persistent Volumes through a CSI-compatible volume plugin in which case  the `CSIPersistentVolume` feature gate must be enabled. | |
```

>  |`ra2.stg.007`|  | An implementation should use Kubernetes Storage Classes to support automation and the separation of concerns between providers of a service and consumers of the service<br>Storage Classes allow a consumer of the Kubernetes platform to request Persistent Storage using a Persistent Volume Claim and for a Persistent Volume to be dynamically created based on the "class" that has been requested. This avoids having to grant `create`/`update`/`delete` permissions in RBAC to PersistentVolume resources, which are cluster-scoped rather than namespace-scoped (meaning an identity can manage all PVs or none) | |

Reviewer 2: Adding final full stop.
```suggestion
 |`ra2.stg.007`|  | An implementation should use Kubernetes Storage Classes to support automation and the separation of concerns between providers of a service and consumers of the service.<br>Storage Classes allow a consumer of the Kubernetes platform to request Persistent Storage using a Persistent Volume Claim and for a Persistent Volume to be dynamically created based on the "class" that has been requested. This avoids having to grant `create`/`update`/`delete` permissions in RBAC to PersistentVolume resources, which are cluster-scoped rather than namespace-scoped (meaning an identity can manage all PVs or none). | |
```

Reviewer 2: I'd also argue that the following part is unnecessary as it should be part of the chapter03.md content.

> Storage Classes allow a consumer of the Kubernetes platform to request Persistent Storage using a Persistent Volume Claim and for a Persistent Volume to be dynamically created based on the "class" that has been requested. This avoids having to grant `create`/`update`/`delete` permissions in RBAC to PersistentVolume resources, which are cluster-scoped rather than namespace-scoped (meaning an identity can manage all PVs or none).
